### PR TITLE
Add ERC721URIStorage extension

### LINKED ISSUE #2554: ERC721TokenUri extension to ERC721
**🧐 Motivation**
see discussion [here](https://forum.openzeppelin.com/t/function-settokenuri-in-erc721-is-gone-with-pragma-0-8-0/5978/8)

**📝 Details**
Add a new extension `ERC721TokenUri` that implements the tokenURI behavior as it was available in 3.4.0. This includes `_setBaseURI` and `_setTokenURI` functions

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #2554

#### PR Checklist
- [x] Tests
- [x] Documentation
- [x] Changelog entry

### CODE REVIEW

**Review by frangio — COMMENTED** [review]
It seems the contract was designed as fully compatible with Contracts 3.x, but I thought the only thing we wanted to preserve was the ability to customize the URI per-token, so only `_setTokenURI`. Defining the base URI through overriding is a better mechanism for gas, and I don't think we lose a significant feature if we remove `_setBaseURI`.

And I would consider defining `tokenURI` so that it delegates to `super.tokenURI` in the fallback case.

I would also remove the public `baseURI`.

Regarding the name, I was thinking something like `ERC721URIStorage`?

File: contracts/mocks/ERC721BurnableMock.sol [hunk 1]
```diff
@@ -7,7 +7,19 @@ import "../token/ERC721/extensions/ERC721Burnable.sol";
 contract ERC721BurnableMock is ERC721Burnable {
     constructor(string memory name, string memory symbol) ERC721(name, symbol) { }
 
+    function exists(uint256 tokenId) public view returns (bool) {
```
[the remark this answers is not in the document]
Comment (Amxx) [reply contracts/mocks/ERC721BurnableMock.sol:10]: Some ERC721 mocks make all internal functions available, other don't. I wanted to improve consistency by making all the mocks show all the functions.

**Review by frangio — APPROVED** [review]

**Comment by sambacha** [comment]
Any chance this will be backported? 

Thanks!

**Comment by Amxx** [comment]
Hello @sambacha,

The ERC721 implementation that is part of previous version of contracts (3.x) does include this by default. We are only adding this as a separate module because this logic was removed from the base ERC721 implementation in 4.0.

What version of contracts are you using, and what exactly do you find missing ?

**Comment by superphly** [comment]
So, how does one implement this contract without having to mutating the other contracts in the repo? I'm currently developing on this 4.0 branch, but I don't want a mutated copy.

**Comment by frangio** [comment]
@superphly I'm not sure I understood your question! But this is an example of how you would use this contract:

```solidity
import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";

contracts MyNFT is ERC721URIStorage {
  constructor() ERC721("Name", "SYM") 
}
```

It's not necessary to add `ERC721` in the inheritance list because it is implied by `ERC721URIStorage`.

**Pull request merged**